File: Blind75-Practice/easy/01-arrays-hashing/arrays/Python/design-hashmap.py
"""
LeetCode 706: Design HashMap
Difficulty: Easy
Pattern: Arrays & Hashing

Implement MyHashMap with put, get, and remove. Keys and values are in [0, 10^6];
get/remove for missing keys returns -1.

Approach: Parallel presence flags and value array — O(1) per op.

Time Complexity: O(1) per operation
Space Complexity: O(10^6)
"""

from typing import List


class MyHashMap:
    def __init__(self) -> None:
        self._has: List[bool] = [False] * 1_000_001
        self._vals: List[int] = [0] * 1_000_001

    def put(self, key: int, value: int) -> None:
        self._has[key] = True
        self._vals[key] = value

    def get(self, key: int) -> int:
        return self._vals[key] if self._has[key] else -1

    def remove(self, key: int) -> None:
        self._has[key] = False


if __name__ == "__main__":
    hm = MyHashMap()
    hm.put(1, 1)
    hm.put(2, 2)
    assert hm.get(1) == 1
    assert hm.get(3) == -1
    hm.put(2, 1)
    assert hm.get(2) == 1
    hm.remove(2)
    assert hm.get(2) == -1
    print("All test cases passed!")

File: Blind75-Practice/easy/01-arrays-hashing/arrays/Java/DesignHashMap.java
/*
706. Design HashMap
Difficulty: Easy
Pattern: Arrays & Hashing

put(key, value), get(key), remove(key). Missing key => get returns -1.
Keys and values in [0, 1_000_000].

Approach: parallel boolean[] for occupancy and int[] for values.

Example:
Input:  ["MyHashMap","put","put","get","get","put","get","remove","get"]
        [[],[1,1],[2,2],[1],[3],[2,1],[2],[2],[2]]
Output: [null,null,null,1,-1,null,1,null,-1]
Explanation:
  put(1,1) → map: {1→1}
  put(2,2) → map: {1→1, 2→2}
  get(1)   → 1
  get(3)   → -1 (not found)
  put(2,1) → map: {1→1, 2→1}  (value updated)
  get(2)   → 1
  remove(2)→ map: {1→1}
  get(2)   → -1 (not found)

Time: O(1) per op, Space: O(10^6)
*/

public class DesignHashMap {

    static class MyHashMap {
            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            //

            // put :
            // has[key] = true;
            // vals[key] = value;

            // get : 
            // return has[key] ? vals[key] : -1;

            // remove:
            // has[key] = false;
        }
    }

    public static void main(String[] args) {
        MyHashMap hm = new MyHashMap();
        hm.put(1, 1);
        hm.put(2, 2);
        assert hm.get(1) == 1;
        assert hm.get(3) == -1;
        hm.put(2, 1);
        assert hm.get(2) == 1;
        hm.remove(2);
        assert hm.get(2) == -1;
        System.out.println("All test cases passed!");
    }
}
